### ISSUE TITLE
Query parameters included in request body for LTI launch requests 

### ISSUE THREAD

**Issue by Tape** [instructure/canvas-lms#600]
Maybe I'm missing something, but I wanted to check in here just to make absolutely sure.

As far as I understand from reading over the spec the POST request is just supposed to contain the launch parameters for the LMS - that's all fine and dandy. However when my launch url contains query parameters, it seems to be signed correctly but Canvas includes the query parameters in the launch url in the request body which throws off the signing process.

For example I have my app configured correctly and the launch url is `https://myapp.com?media=51`, `media=51` is included in the request body. When I remove these parameters from the request body and generate the signature per the spec everything works great. However if I leave everything untouched `media=51` appears in the signature twice because the request body has it (unexpected) and the query parameters have it (expected).

Let me know if more information is needed, but this appears to happen both with `course_navigation` and `editor_button` from what I can tell.

Thanks!

**Comment by Tape** [comment]
I think [this](https://github.com/instructure/canvas-lms/blob/10319c8df4a4b02ab60b059c46448c2df34a1170/gems/lti_outbound/lib/lti_outbound/tool_launch.rb#L180-L205) may be the offending block of code. Basically the get parameters are always added to the front-end form when they should only be used for generating the signed request.

**Comment by wbhumphrey** [comment]
You are correct.  The reason that Canvas does this is because we needed a way to add link level custom variables to launches, and iterating over a link and add GET parameters to the POST body was a way to solve it.  This shouldn't cause a error when verifying the signature though, because you should be creating the signature base string from all of the GET and POST params that are sent to you on launch.  This would be important for a robust implementation even if canvas didn't replicate GET params in the POST body, because it is possible for a tool consumer to send parameters that matched supplied GET parameters inadvertently.

**Comment by Tape** [comment]
Just to be clear, would your suggested implementation be to ignore exact duplicates from POST and GET? You are absolutely correct the signature comes out as expected, just the GET parameters appearing again in the POST throw a wrench in the situation. I help maintain the extension I'm using so if this is an understood issue I'm more than happy to fix the bug.

**Comment by wbhumphrey** [comment]
According to the spec, the way you are building the string is actually correct.

>The parameters are sorted by name, using ascending byte value
>ordering.  If two or more parameters share the same name, they
>are sorted by their value.

It looks like there are a couple problems in Canvas.  The first is that the ruby oauth library we are using seems to ignore duplicate key/value pairs sometimes (maybe when they are identical; I am not sure).  The second is that canvas is duplicating variables from query string to POST in a bad way.  In fact, while testing this, I found that there are situations where canvas is deleting valid duplicate values (ie query string arrays).

The tricky part is changing they way canvas signs, because it could break currently working implementations, so it isn't something we can change hastily.  I think the most robust solution will be something that tries to reduce the number of duplicate params sent automatically by canvas.

My recommendation is that you do not change your OAuth 1 signing library, but instead implement a short term fix specific to canvas implementations if it is possible.  Hopefully in the next month or two will will be able to roll out a fix to the way we are signing to be in better alignment with the spec.

**Comment by bhecht2** [comment]
Hi,

This is sad news.
I have an app working for all the different LMS's around there, but cannot work with Canvas.
From what i understand, Canvas signs wrongly an LTI launch URL that includes parameters in it.
All other LMS's sign this fine.
In addition you are saying Canvas will not be fixing this?
What are my options?

**Comment by rivernate** [comment]
@bhecht2 we have added code to fix the signing problems in these two commits: ad2a091 and 5dcf89b
It does require an admin to turn on the feature flag "Don't Move LTI Query Params to POST Body" though.  It's something we want to make the default eventually we just have to make sure that we don't break existing installs/tools

**Comment by dlants** [comment]
Could you clarify how canvas generates signatures for request parameters?

In particular, in this flow http://lti.tools/oauth/ I set:
`POST` request
path to `lti/launch?param=true`
params to `param=true`
![screen shot 2017-04-14 at 11 18 58 am](https://cloud.githubusercontent.com/assets/1186977/25051963/355c5574-2104-11e7-8ecb-61af58880c34.png)

The base string that's generated is this:

```
POST&http%3A%2F%2Flti.com%2Flti%2Flaunch%3Fparam%3Dtrue&oauth_consumer_key%3D%26oauth_nonce%3D%26oauth_signature_method%3DHMAC-SHA1%26oauth_timestamp%3D0%26oauth_version%3D1.0%26param%3Dtrue
```

What would canvas do in this situation?

**Comment by ziaulrehman40** [comment]
I just faced same issue of signature(i think its signature issue, ruby's lms-lti gem just fails on `@provider.valid_request?(request)`)

Everything works if there are no custom params, we are building a cross LMS tool, we implented moodle, it provides specific separate fields which specify custom LTI params which we want to send along.
But in canvas, i first had to guess that lets use get type params, it worked, but than i am facing this issue, i think its the same issue. Any help will be appreciated, thanks.
